test/integration/replicaset/replicaset_test.go

```diff
@@ -515,3 +519,682 @@ func TestRSSelectorImmutability(t *testing.T) {
+	}
+	// Verify PATCH erases OwnerReferences of pod2 as the RS can only have 1 pod
+	if pod2.OwnerReferences != nil {
+		t.Fatalf("Unexpected ControllerReference for pod %s: Expected nil, saw %v", pod2.Name, pod.OwnerReferences)
+	}
+
+	// Should we test when the pod belongs to another RS? It seems irrelevant.
+}
+
+// addPodConditionReady sets given pod status to ready at given time
+func addPodConditionReady(pod *v1.Pod, time metav1.Time) {
+	pod.Status = v1.PodStatus{
+		Phase: v1.PodRunning,
+		Conditions: []v1.PodCondition{
+			{
+				Type:               v1.PodReady,
+				Status:             v1.ConditionTrue,
+				LastTransitionTime: time,
+			},
+		},
+	}
+}
+
+func markAllPodsReady(t *testing.T, clientSet clientset.Interface, rs *v1beta1.ReplicaSet, ns string) {
+	selector, err := metav1.LabelSelectorAsSelector(rs.Spec.Selector)
+	if err != nil {
+		t.Fatalf("failed to parse ReplicaSet selector: %v", err)
+	}
+	var readyPods int32
+	err = wait.Poll(1*time.Second, 1*time.Minute, func() (bool, error) {
+		readyPods = 0
+		pods, err := clientSet.Core().Pods(ns).List(metav1.ListOptions{LabelSelector: selector.String()})
+		if err != nil {
+			t.Logf("failed to list ReplicaSet pods, will retry later: %v", err)
+			return false, nil
+		}
+		for i := range pods.Items {
+			pod := pods.Items[i]
+			if podutil.IsPodReady(&pod) {
+				readyPods++
+				continue
+			}
+			addPodConditionReady(&pod, metav1.Now())
+			if _, err = clientSet.Core().Pods(ns).UpdateStatus(&pod); err != nil {
+				t.Logf("failed to update ReplicaSet pod %s, will retry later: %v", pod.Name, err)
+			} else {
+				readyPods++
+			}
+		}
+		if readyPods >= *rs.Spec.Replicas {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to mark all ReplicaSet pods to ready: %v", err)
+	}
+	//fmt.Println("readyPods: %d, rs.spec.replicas: %d", readyPods, *rs.Spec.Replicas)
+}
+
+func waitAllPodsReady(t *testing.T, clientSet clientset.Interface, rs *v1beta1.ReplicaSet, ns string) {
+	rsClient := clientSet.Extensions().ReplicaSets(ns)
+	if err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		updatedRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if updatedRS.Status.ReadyReplicas != *rs.Spec.Replicas {
+			return false, nil
+		} else {
+			return true, nil
+		}
+	}); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestReadyAndAvailableReplicas(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-ready-and-available-replicas", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 3)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	// Make .spec.minReadySeconds long enough so that we can test the replicas are unavailable initially
+	rs.Spec.MinReadySeconds = 3600
+
+	// Create 3 pods
```

Comment:
Let the RS create the Pods. Set Pods as Running also in markAllPodsReady.

Doing it like this would be racy, because you create the RS and Pods asynchronously. You don't know whether the Pods that eventually survive are the ones you created or ones created by RS before you got to it.

Reply:
Yup! I am not aware that there is a possibility that the pods that eventually survive may not be the ones that i created. Thanks for pointing the fact out!